Core/src/main/java/org/tribuo/protos/ProtoUtil.java

```diff
@@ -0,0 +1,474 @@
+                        valueAdder.setAccessible(true);
+                        Map<?,?> map = (Map<?,?>) obj;
+                        if (map != null) {
+                            for (Map.Entry<?,?> e : map.entrySet()) {
+                                keyAdder.invoke(serializedDataBuilder, convert(e.getKey()));
+                                valueAdder.invoke(serializedDataBuilder, convert(e.getValue()));
+                            }
+                        }
+                        keyAdder.setAccessible(false);
+                        valueAdder.setAccessible(false);
+                    }
+
+                    ProtoSerializableMapValuesField psmvf = field.getAnnotation(ProtoSerializableMapValuesField.class);
+                    if (psmvf != null) {
+                        if (!(obj instanceof Map) && (obj != null)) {
+                            throw new IllegalStateException("Field of type " + obj.getClass() + " was annotated with ProtoSerializableMapValuesField which is only supported on java.util.Map fields");
+                        }
+                        Method valuesAdder = findMethod(serializedDataBuilderClass, "addAll", psmvf.valuesName(), 1);
+                        obj = convertValues((Map<?,?>) obj);
+                        valuesAdder.setAccessible(true);
+                        valuesAdder.invoke(serializedDataBuilder, obj);
+                        valuesAdder.setAccessible(false);
+                    }
+
+                    field.setAccessible(false);
+                }
+                serializedClassBuilderClass.getMethod("setSerializedData", com.google.protobuf.Any.class).invoke(serializedClassBuilder, Any.pack(serializedDataBuilder.build()));
+            }
+            return (SERIALIZED_CLASS) serializedClassBuilder.build();
+        } catch (InvocationTargetException | IllegalAccessException | NoSuchMethodException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * If this class is annotated with {@link ProtoSerializableClass} returns the
+     * version number, otherwise returns -1.
+     *
+     * @param clazz The class to check.
+     * @return The version number, or -1 if it does not use the annotation.
+     */
+    public static int getCurrentVersion(Class<? extends ProtoSerializable<?>> clazz) {
+        ProtoSerializableClass ann = clazz.getAnnotation(ProtoSerializableClass.class);
+        if (ann != null) {
+            return ann.version();
+        } else {
+            return -1;
+        }
+    }
+
+    static <SERIALIZED_CLASS extends Message, PROTO_SERIALIZABLE extends ProtoSerializable<SERIALIZED_CLASS>> Class<SERIALIZED_CLASS> getSerializedClass(PROTO_SERIALIZABLE protoSerializable) {
+        @SuppressWarnings("unchecked")
+        Class<SERIALIZED_CLASS> serializedClass = (Class<SERIALIZED_CLASS>) resolveTypeParameter(ProtoSerializable.class, protoSerializable.getClass(), ProtoSerializable.class.getTypeParameters()[0]);
```

Comment:
If the concrete `protoSerializable` has its own type parameters, does this still work properly?

Reply:
I'll check.

Reply:
I think this is tested in `ProtoUtilTest.testGetSerializedClass()`.